source/pervasive/nonlinear_arith/div.rs

```diff
@@ -0,0 +1,1088 @@
+#[allow(unused_imports)]
+use crate::nonlinear_arith::internals::div_internals_nonlinear as DivINL;
+#[allow(unused_imports)]
+use crate::nonlinear_arith::internals::mod_internals::{lemma_mod_auto, mod_recursive};
+#[allow(unused_imports)]
+use crate::nonlinear_arith::internals::mod_internals_nonlinear as ModINL;
+#[allow(unused_imports)]
+use crate::nonlinear_arith::internals::mul_internals::{lemma_mul_induction_auto, lemma_mul_auto, mul_auto, lemma_mul_induction};
+#[allow(unused_imports)]
+use crate::nonlinear_arith::mul::*;
+#[allow(unused_imports)]
+use crate::nonlinear_arith::internals::general_internals::{is_le};
+#[allow(unused_imports)]
+use crate::nonlinear_arith::modulus::*;
+
+/*****************************************************************************
+ * Division:
+ *****************************************************************************/
+
+/// the common syntax of division gives the same quotient as performing division through recursion
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_is_div_recursive(x: int, d: int)
+    requires 0 < d
+    ensures div_recursive(x, d) == x / d
+{
+    reveal(div_recursive);
+    reveal(div_pos);
+    lemma_div_induction_auto(d, x, |u: int| div_recursive(u, d) == u / d);
+}
+
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_is_div_recursive_auto()
+    ensures forall |x: int, d: int| d > 0 ==> div_recursive(x, d) == #[trigger](x / d)
+{
+    reveal(div_recursive);
+    assert forall |x: int, d: int| d > 0 implies div_recursive(x, d) == #[trigger](x / d) by
+    {
+    lemma_div_is_div_recursive(x, d);
+    }
+}
+
+/// the quotient of an integer divided by itself is 1
+// #[verifier::spinoff_prover] 
+pub proof fn lemma_div_by_self(d: int)
+    requires d != 0
+    ensures d / d == 1
+{
+    DivINL::lemma_div_by_self(d);
+}
+
+/// zero divided by an integer besides 0 is 0
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_of0(d: int)
+    requires d != 0
+    ensures 0 as int  / d == 0
+{
+    DivINL::lemma_div_of0(d);
+}
+
+/// ensures the basic propoerties of division: 0 divided by any integer is 0; any integer 
+/// divided by 1 is itself; any integer divided by itself is 1
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_basics(x: int)
+    ensures 
+        x != 0 as int ==> 0 as int / x == 0,
+        x / 1 == x,
+        x != 0 ==> x / x == 1,
+{
+    if (x != 0) {
+    lemma_div_by_self(x);
+    lemma_div_of0(x);
+    }
+}
+
+// #[verifier::spinoff_prover]
+pub proof fn lemma_div_basics_auto()
+    ensures
+        forall |x: int| x != 0 ==> #[trigger](0int / x) == 0,
+        forall |x: int| #[trigger](x / 1) == x,
+        forall |x: int, y: int| x >= 0 && y > 0 ==> #[trigger](x / y) >= 0,
+        forall |x: int, y: int| x >= 0 && y > 0 ==> #[trigger](x / y) <= x,
+{
+    assert forall |x: int| (x != 0 ==> #[trigger](0int / x) / x == 0) && (#[trigger](x / 1) == x) by
+    {
+    lemma_div_basics(x);
+    };
```

Comment:
Hmm, there appear to be indentation issues of this kind scattered throughout the codebase, might be good to look them over and fix them up